
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Genomics Aotearoa" name="author"/>
<link href="https://genomicsaotearoa.github.io/Introduction-to-R/04-bioconductor-vcfr/" rel="canonical"/>
<link href="../03-basics-factors-dataframes/" rel="prev"/>
<link href="../05-dplyr/" rel="next"/>
<link href="../theme_figures/nesi_ga.png" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.0.13" name="generator"/>
<title>Using packages from Bioconductor - Introduction to R</title>
<link href="../assets/stylesheets/main.ffa9267a.min.css" rel="stylesheet"/>
<link href="../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
<link href="../stylesheets/extra.css" rel="stylesheet"/>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="indigo" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#using-packages-from-bioconductor">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="Introduction to R" class="md-header__button md-logo" data-md-component="logo" href=".." title="Introduction to R">
<img alt="logo" src="../theme_figures/nesi_ga.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            Introduction to R
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Using packages from Bioconductor
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme)" data-md-color-primary="indigo" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_3" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"></path></svg>
</label>
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="indigo" data-md-color-scheme="default" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"></path></svg>
</label>
<input aria-label="Switch to system preference" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="indigo" data-md-color-scheme="slate" id="__palette_3" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to system preference">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"></path></svg>
</label>
</form>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="Introduction to R" class="md-nav__button md-logo" data-md-component="logo" href=".." title="Introduction to R">
<img alt="logo" src="../theme_figures/nesi_ga.png"/>
</a>
    Introduction to R
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
        Home
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../00-introduction/">
        Introducing R and RStudio IDE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../01-r-basics/">
        R Basics
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../02-data-prelude/">
        Introduction to the example dataset and file type
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../03-basics-factors-dataframes/">
        R Basics continued - factors and data frames
      </a>
</li>
<li class="md-nav__item md-nav__item--active">
<input class="md-nav__toggle md-toggle" id="__toc" type="checkbox"/>
<label class="md-nav__link md-nav__link--active" for="__toc">
          Using packages from Bioconductor
          <span class="md-nav__icon md-icon"></span>
</label>
<a class="md-nav__link md-nav__link--active" href="./">
        Using packages from Bioconductor
      </a>
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#installing-packages-from-somewhere-else-besides-cran">
    Installing packages from somewhere else besides CRAN?
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#so-how-do-i-use-it">
    So how do I use it?
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#first-install-the-biocmanager-package">
    First, install the BiocManager package
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager">
    Second, install the vcfR package from Bioconductor using BiocManager
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies">
    Head's Up: Installing vcfR may take a while due to numerous dependencies
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#note-installing-packages-from-bioconductor-vs-from-cran">
    Note: Installing packages from Bioconductor vs from CRAN
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#search-for-bioconductor-packages-based-on-your-analysis-needs">
    Search for Bioconductor packages based on your analysis needs
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#tip-searching-for-packages-on-the-bioconductor-website">
    Tip: Searching for packages on the Bioconductor website
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#challenge">
    Challenge
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#tip-refreshing-the-rstudio-package-view-after-installing">
    Tip: Refreshing the RStudio package view after installing
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#resources">
    Resources
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../05-dplyr/">
        Data Wrangling and Analyses with Tidyverse
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../06-data-visualization/">
        Data Visualization with ggplot2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../07-knitr-markdown/">
        Producing Reports With knitr
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../08-r-help/">
        Getting help with R
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#installing-packages-from-somewhere-else-besides-cran">
    Installing packages from somewhere else besides CRAN?
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#so-how-do-i-use-it">
    So how do I use it?
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#first-install-the-biocmanager-package">
    First, install the BiocManager package
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager">
    Second, install the vcfR package from Bioconductor using BiocManager
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies">
    Head's Up: Installing vcfR may take a while due to numerous dependencies
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#note-installing-packages-from-bioconductor-vs-from-cran">
    Note: Installing packages from Bioconductor vs from CRAN
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#search-for-bioconductor-packages-based-on-your-analysis-needs">
    Search for Bioconductor packages based on your analysis needs
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#tip-searching-for-packages-on-the-bioconductor-website">
    Tip: Searching for packages on the Bioconductor website
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#challenge">
    Challenge
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#tip-refreshing-the-rstudio-package-view-after-installing">
    Tip: Refreshing the RStudio package view after installing
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#resources">
    Resources
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="using-packages-from-bioconductor">Using packages from Bioconductor<a class="headerlink" href="#using-packages-from-bioconductor" title="Permanent link">¶</a></h1>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>keypoints:</p>
<ul>
<li>Bioconductor is an alternative package repository for bioinformatics
  packages.</li>
<li>Installing packages from Bioconductor requires a new method, since it
  is not compatible with the <code>install.packages()</code> function used for
  CRAN.</li>
<li>Check Bioconductor to see if there is a package relevant to your
  analysis before writing code yourself.</li>
</ul>
<p>objectives:</p>
<ul>
<li>Describe what the Bioconductor repository is and what it is used for</li>
<li>Describe how Bioconductor differs from CRAN</li>
<li>Search Bioconductor for relevant packages</li>
<li>Install a package from Bioconductor</li>
</ul>
<p>questions:</p>
<ul>
<li>How do I use packages from the Bioconductor repository?</li>
</ul>
</div>
<h2 id="installing-packages-from-somewhere-else-besides-cran">Installing packages from somewhere else besides CRAN?<a class="headerlink" href="#installing-packages-from-somewhere-else-besides-cran" title="Permanent link">¶</a></h2>
<p>So far we have told you about using packages that are included in the
base installation of R (this is what comes with R 'out of the box'), and
packages that you can install from <a href="https://cran.r-project.org/">CRAN</a>
(the Comprehensive R Archive Network), which is the primary place many
people look for supplemental R packages to install. However, not all R
packages are available on CRAN. For bioinformatics-related packages in
particular, there is another repository that has many powerful packages
that you can install. It is called
<a href="https://bioconductor.org/">Bioconductor</a> and it is a repository
specifically focused on bioinformatics packages.
<a href="https://bioconductor.org/">Bioconductor</a> has a mission of
"promot[ing] the statistical analysis and comprehension of current and
emerging high-throughput biological assays." This means that many if not
all of the packages available on Bioconductor are focused on the
analysis of biological data, and that it can be a great place to look
for tools to help you analyze your -omics datasets!</p>
<h2 id="so-how-do-i-use-it">So how do I use it?<a class="headerlink" href="#so-how-do-i-use-it" title="Permanent link">¶</a></h2>
<p>Since access to the <a href="https://bioconductor.org/">Bioconductor</a> repository
is not built in to base R 'out of the box', there are a couple steps
needed to install packages from this alternative source. We will work
through the steps (only 2!) to install a package to help with the VCF
analysis we are working on, but you can use the same approach to install
any of the many thousands of available packages.</p>
<p>[screenshot of bioconductor homepage]{.image .placeholder
original-image-src="../fig/bioconductor_website_screenshot.jpg"
original-image-title="fig:"}</p>
<h2 id="first-install-the-biocmanager-package">First, install the <code>BiocManager</code> package<a class="headerlink" href="#first-install-the-biocmanager-package" title="Permanent link">¶</a></h2>
<p>The first step is to install a package that <em>is</em> on CRAN, <code>BiocManager</code>.
This package will allow us to use it to install packages from
Bioconductor. You can think of Bioconductor kind of like an alternative
app store for your phone, except instead of apps you are installing
packages, and instead of your phone it's your local R package library.</p>
<p><code>{r install-biocmanager, eval = FALSE} # install the BiocManager from CRAN using the base R install.packages() function install.packages("BiocManager")</code></p>
<p>To check if this worked (and also so you can make a note of the version
for reproducibility purposes), you can run <code>BiocManager::version()</code> and
it should give you the version number.</p>
<p><code>{r check-bioconductor-version, eval = FALSE} # to make sure it worked, check the version BiocManager::version()</code></p>
<h2 id="second-install-the-vcfr-package-from-bioconductor-using-biocmanager">Second, install the vcfR package from Bioconductor using <code>BiocManager</code><a class="headerlink" href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager" title="Permanent link">¶</a></h2>
<blockquote>
<h2 id="heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies">Head's Up: Installing vcfR may take a while due to numerous dependencies<a class="headerlink" href="#heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies" title="Permanent link">¶</a></h2>
<p>Just be aware that installing packages that have many dependencies can
take a while.</p>
<p>{: .callout}</p>
</blockquote>
<p><code>{r install-vcfR, eval = FALSE} # install the vcfR package from bioconductor using BiocManager::install() BiocManager::install("vcfR")</code>
Depending on your particular system, you may need to also allow it to
install some dependencies or update installed packages in order to
successfully complete the process.</p>
<blockquote>
<h2 id="note-installing-packages-from-bioconductor-vs-from-cran">Note: Installing packages from Bioconductor vs from CRAN<a class="headerlink" href="#note-installing-packages-from-bioconductor-vs-from-cran" title="Permanent link">¶</a></h2>
<p>Some packages begin by being available only on Bioconductor, and then
later move to CRAN. <code>vcfR</code> is one such package, which originally was
only available from Bioconductor, but is currently available from
CRAN. The other thing to know is that <code>BiocManager::install()</code> will
also install packages from CRAN (it is a wrapper around
<code>install.packages()</code> that adds some extra features). There are other
benefits to using <code>BiocManager::install()</code> for Bioconductor packages,
many of which are outlined
<a href="https://www.bioconductor.org/install/">here</a>. In short, Bioconductor
packages have a release cycle that is different from CRAN and the
<code>install()</code> function is aware of that difference, so it helps to keep
package versions in line with one another in a way that doesn't
generally happen with the base R <code>install.packages()</code>.</p>
<p>{: .callout}</p>
</blockquote>
<h2 id="search-for-bioconductor-packages-based-on-your-analysis-needs">Search for Bioconductor packages based on your analysis needs<a class="headerlink" href="#search-for-bioconductor-packages-based-on-your-analysis-needs" title="Permanent link">¶</a></h2>
<p>While we are only focusing in this workshop on VCF analyses, there are
hundreds or thousands of different types of data and analyses that
bioinformaticians may want to work with. Sometimes you may get a new
dataset and not know exactly where to start with analyzing or
visualizing it. The Bioconductor package search view can be a great way
to browse through the packages that are available.</p>
<p>[screenshot of bioconductor search]{.image .placeholder
original-image-src="../fig/bioconductor_search.jpg"
original-image-title="fig:"}</p>
<blockquote>
<h2 id="tip-searching-for-packages-on-the-bioconductor-website">Tip: Searching for packages on the Bioconductor website<a class="headerlink" href="#tip-searching-for-packages-on-the-bioconductor-website" title="Permanent link">¶</a></h2>
<p>There are several thousand packages available through the Bioconductor
website. It can be a bit of a challenge to find what you want, but one
helpful resource is the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">package search
page</a>.</p>
<p>{: .callout}</p>
</blockquote>
<p>In bioinformatics, there are often many different tools that can be used
in a particular instance. The authors of <code>vcfR</code> have <a href="https://github.com/knausb/vcfR#software-that-produce-vcf-files">compiled some of
them</a>.
One of those packages that <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html">is available from
Bioconductor</a>
is called <code>VariantAnnotation</code> and may also be of interest to those
working with vcf files in R.</p>
<blockquote>
<h2 id="challenge">Challenge<a class="headerlink" href="#challenge" title="Permanent link">¶</a></h2>
<ul>
<li>Use the <code>BiocManager::available()</code> function to see what packages
    are available matching a search term.</li>
<li>Use the
    <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">biocViews</a>
    interface to search for packages of interest.</li>
</ul>
<p>You may or may not want to try installing the package, since not all
dependencies always install easily. However, this will at least let
you see what is available. {: .challenge}</p>
<h2 id="tip-refreshing-the-rstudio-package-view-after-installing">Tip: Refreshing the RStudio package view after installing<a class="headerlink" href="#tip-refreshing-the-rstudio-package-view-after-installing" title="Permanent link">¶</a></h2>
<p>If you install a package from Bioconductor, you may need to refresh
the RStudio package view to see it in your list. You can do this by
clicking the "Refresh" button in the Packages pane of RStudio.</p>
<p>{: .callout}</p>
</blockquote>
<h2 id="resources">Resources<a class="headerlink" href="#resources" title="Permanent link">¶</a></h2>
<ul>
<li><a href="https://bioconductor.org/">Bioconductor</a></li>
<li><a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">Bioconductor package
    search</a></li>
<li><a href="https://cran.r-project.org/">CRAN</a></li>
</ul>
</article>
</div>
</div>
<a class="md-top md-icon" data-md-component="top" hidden="" href="#">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"></path></svg>
            Back to top
          </a>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top"], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../assets/javascripts/bundle.2a6f1dda.min.js"></script>
<script src="../javascripts/mathjax.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>