
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Genomics Aotearoa" name="author"/>
<link href="https://genomicsaotearoa.github.io/Introduction-to-R/01-r-basics/" rel="canonical"/>
<link href="../00-introduction/" rel="prev"/>
<link href="../02-data-prelude/" rel="next"/>
<link href="../theme_figures/nesi_ga.png" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.0.13" name="generator"/>
<title>R Basics - Introduction to R</title>
<link href="../assets/stylesheets/main.ffa9267a.min.css" rel="stylesheet"/>
<link href="../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style>
<link href="../stylesheets/extra.css" rel="stylesheet"/>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="indigo" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#r-basics">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="Introduction to R" class="md-header__button md-logo" data-md-component="logo" href=".." title="Introduction to R">
<img alt="logo" src="../theme_figures/nesi_ga.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            Introduction to R
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              R Basics
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme)" data-md-color-primary="indigo" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_3" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"></path></svg>
</label>
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="indigo" data-md-color-scheme="default" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"></path></svg>
</label>
<input aria-label="Switch to system preference" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="indigo" data-md-color-scheme="slate" id="__palette_3" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to system preference">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"></path></svg>
</label>
</form>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="Introduction to R" class="md-nav__button md-logo" data-md-component="logo" href=".." title="Introduction to R">
<img alt="logo" src="../theme_figures/nesi_ga.png"/>
</a>
    Introduction to R
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/GenomicsAotearoa/Introduction-to-R" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GenomicsAotearoa/Introduction-to-R
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
        Home
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../00-introduction/">
        Introducing R and RStudio IDE
      </a>
</li>
<li class="md-nav__item md-nav__item--active">
<input class="md-nav__toggle md-toggle" id="__toc" type="checkbox"/>
<label class="md-nav__link md-nav__link--active" for="__toc">
          R Basics
          <span class="md-nav__icon md-icon"></span>
</label>
<a class="md-nav__link md-nav__link--active" href="./">
        R Basics
      </a>
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r">
    "The fantastic world of R awaits you" OR "Nobody wants to learn how to use R"
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#what-these-lessons-will-not-teach-you">
    What these lessons will not teach you
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#creating-objects-in-r">
    Creating objects in R
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#naming-objects-in-r">
    Naming objects in R
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#reassigning-object-names-or-deleting-objects">
    Reassigning object names or deleting objects
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#understanding-object-data-types-modes">
    Understanding object data types (modes)
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mathematical-and-functional-operations-on-objects">
    Mathematical and functional operations on objects
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#creating-and-subsetting-vectors">
    Creating and subsetting vectors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#adding-to-removing-or-replacing-values-in-existing-vectors">
    Adding to, removing, or replacing values in existing vectors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#exercise-examining-and-subsetting-vectors">
    Exercise: Examining and subsetting vectors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#logical-subsetting">
    Logical Subsetting
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#the-magic-of-programming">
    The magic of programming
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#a-few-final-vector-tricks">
    A few final vector tricks
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-1">
    Review Exercise 1
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-2">
    Review Exercise 2
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_1">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-3">
    Review Exercise 3
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_2">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-4">
    Review Exercise 4
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_3">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-5">
    Review Exercise 5
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_4">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#bonus-material-lists">
    Bonus material: Lists
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../02-data-prelude/">
        Introduction to the example dataset and file type
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../03-basics-factors-dataframes/">
        R Basics continued - factors and data frames
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../04-bioconductor-vcfr/">
        Using packages from Bioconductor
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../05-dplyr/">
        Data Wrangling and Analyses with Tidyverse
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../06-data-visualization/">
        Data Visualization with ggplot2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../07-knitr-markdown/">
        Producing Reports With knitr
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../08-r-help/">
        Getting help with R
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r">
    "The fantastic world of R awaits you" OR "Nobody wants to learn how to use R"
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#what-these-lessons-will-not-teach-you">
    What these lessons will not teach you
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#creating-objects-in-r">
    Creating objects in R
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#naming-objects-in-r">
    Naming objects in R
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#reassigning-object-names-or-deleting-objects">
    Reassigning object names or deleting objects
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#understanding-object-data-types-modes">
    Understanding object data types (modes)
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mathematical-and-functional-operations-on-objects">
    Mathematical and functional operations on objects
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#creating-and-subsetting-vectors">
    Creating and subsetting vectors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#adding-to-removing-or-replacing-values-in-existing-vectors">
    Adding to, removing, or replacing values in existing vectors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#exercise-examining-and-subsetting-vectors">
    Exercise: Examining and subsetting vectors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#logical-subsetting">
    Logical Subsetting
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#the-magic-of-programming">
    The magic of programming
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#a-few-final-vector-tricks">
    A few final vector tricks
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-1">
    Review Exercise 1
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-2">
    Review Exercise 2
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_1">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-3">
    Review Exercise 3
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_2">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-4">
    Review Exercise 4
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_3">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#review-exercise-5">
    Review Exercise 5
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#solution_4">
    Solution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#bonus-material-lists">
    Bonus material: Lists
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="r-basics">R Basics<a class="headerlink" href="#r-basics" title="Permanent link">¶</a></h1>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>keypoints:</p>
<ul>
<li>Effectively using R is a journey of months or years. Still you don't
  have to be an expert to use R and you can start using and analyzing
  your data with with about a day's worth of training</li>
<li>It is important to understand how data are organized by R in a given
  object type and how the mode of that type (e.g. numeric, character,
  logical, etc.) will determine how R will operate on that data.</li>
<li>Working with vectors effectively prepares you for understanding how
  data are organized in R.</li>
</ul>
<p>objectives:</p>
<ul>
<li>Be able to create the most common R objects including vectors</li>
<li>Understand that vectors have modes, which correspond to the type of
  data they contain</li>
<li>Be able to use arithmetic operators on R objects</li>
<li>Be able to retrieve (subset), name, or replace, values from a vector</li>
<li>Be able to use logical operators in a subsetting operation</li>
<li>Understand that lists can hold data of more than one mode and can be
  indexed</li>
</ul>
<p>questions:</p>
<ul>
<li>What will these lessons not cover?</li>
<li>What are the basic features of the R language?</li>
<li>What are the most common objects in R?</li>
</ul>
</div>
<h2 id="the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r">"The fantastic world of R awaits you" OR "Nobody wants to learn how to use R"<a class="headerlink" href="#the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r" title="Permanent link">¶</a></h2>
<p>Before we begin this lesson, we want you to be clear on the goal of the
workshop and these lessons. We believe that every learner can <strong>achieve
competency with R</strong>. You have reached competency when you find that you
are able to <strong>use R to handle common analysis challenges in a reasonable
amount of time</strong> (which includes time needed to look at learning
materials, search for answers online, and ask colleagues for help). As
you spend more time using R (there is no substitute for regular use and
practice) you will find yourself gaining competency and even expertise.
The more familiar you get, the more complex the analyses you will be
able to carry out, with less frustration, and in less time - the
fantastic world of R awaits you!</p>
<h2 id="what-these-lessons-will-not-teach-you">What these lessons will not teach you<a class="headerlink" href="#what-these-lessons-will-not-teach-you" title="Permanent link">¶</a></h2>
<p>Nobody wants to learn how to use R. People want to learn how to use R to
analyze their own research questions! Ok, maybe some folks learn R for
R's sake, but these lessons assume that you want to start analyzing
genomic data as soon as possible. Given this, there are many valuable
pieces of information about R that we simply won't have time to cover.
Hopefully, we will clear the hurdle of giving you just enough knowledge
to be dangerous, which can be a high bar in R! We suggest you look into
the additional learning materials in the tip box below.</p>
<p><strong>Here are some R skills we will <em>not</em> cover in these lessons</strong></p>
<ul>
<li>How to create and work with R matrices and R lists</li>
<li>How to create and work with loops and conditional statements, and
    the "apply" family of functions (which are super useful, read more
    <a href="https://www.r-bloggers.com/r-tutorial-on-the-apply-family-of-functions/">here</a>)</li>
<li>How to do basic string manipulations (e.g. finding patterns in text
    using grep, replacing text)</li>
<li>How to plot using the default R graphic tools (we <em>will</em> cover plot
    creation, but will do so using the popular plotting package
    <code>ggplot2</code>)</li>
<li>How to use advanced R statistical functions</li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Where to learn more</p>
<p>The following are good resources for learning more about R. Some of
them can be quite technical, but if you are a regular R user you may
ultimately need this technical knowledge. - <a href="https://cran.r-project.org/doc/contrib/Paradis-rdebuts_en.pdf">R for
Beginners</a>:
By Emmanuel Paradis and a great starting point - <a href="https://cran.r-project.org/manuals.html">The R
Manuals</a>: Maintained by the R
project - <a href="https://cran.r-project.org/other-docs.html">R contributed
documentation</a>: Also
linked to the R project; importantly there are materials available in
several languages - <a href="http://r4ds.had.co.nz/">R for Data Science</a>: A
wonderful collection by noted R educators and developers Garrett
Grolemund and Hadley Wickham - <a href="https://peerj.com/collections/50-practicaldatascistats/">Practical Data Science for
Stats</a>: Not
exclusively about R usage, but a nice collection of pre-prints on data
science and applications for R - <a href="https://software-carpentry.org/lessons/">Programming in R Software Carpentry
lesson</a>: There are several
Software Carpentry lessons in R to choose from </p>
</div>
<h2 id="creating-objects-in-r">Creating objects in R<a class="headerlink" href="#creating-objects-in-r" title="Permanent link">¶</a></h2>
<div class="admonition bell">
<p class="admonition-title">Reminder</p>
<p>At this point you should be coding along in the
"<strong>genomics_r_basics.R</strong>" script we created in the last episode.
Writing your commands in the script (and commenting it) will make it
easier to record what you did and why. </p>
</div>
<p>What might be called a variable in many languages is called an
<strong>object</strong> in R.</p>
<p><strong>To create an object you need:</strong></p>
<ul>
<li>a name (e.g. 'a')</li>
<li>a value (e.g. '1')</li>
<li>the assignment operator ('\&lt;-')</li>
</ul>
<p>In your script, "<strong>genomics_r_basics.R</strong>", using the R assignment
operator '\&lt;-', assign '1' to the object 'a' as shown. Remember to leave
a comment in the line above (using the '#') to explain what you are
doing:</p>
<div class="admonition r-project">
<p class="admonition-title">r</p>
<div class="highlight"><pre><span></span><code><a href="#__codelineno-0-1" id="__codelineno-0-1" name="__codelineno-0-1"></a># this line creates the object 'a' and assigns it the value '1'
<a href="#__codelineno-0-2" id="__codelineno-0-2" name="__codelineno-0-2"></a>
<a href="#__codelineno-0-3" id="__codelineno-0-3" name="__codelineno-0-3"></a>a \&lt;- 1
</code></pre></div>
</div>
<p>Next, run this line of code in your script. You can run a line of code
  by hitting the <code>Run</code> button that is just above the first line of your
  script in the header of the Source pane or you can use the appropriate shortcut:</p>
<ul>
<li>Windows execution shortcut: <code>Ctrl+Enter</code></li>
<li>
<p>Mac execution shortcut: <code>Cmd(⌘)+Enter</code>
  To run multiple lines of code, you can highlight all the line you wish to run
  and then hit Run or use the shortcut key combo listed above.</p>
<p>In the RStudio 'Console' you should see:</p>
<p class="output"><div class="highlight"><pre><span></span><code><a href="#__codelineno-1-1" id="__codelineno-1-1" name="__codelineno-1-1"></a>a &lt;- 1
<a href="#__codelineno-1-2" id="__codelineno-1-2" name="__codelineno-1-2"></a>&gt;
</code></pre></div></p>
<p>The 'Console' will display lines of code run from a script and any outputs or
status/warning/error messages (usually in red).</p>
<p>In the 'Environment' window you will also get a table:</p>
<table>
<thead>
<tr>
<th>Values</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>a</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>The 'Environment' window allows you to keep track of the objects you have
created in R.</p>
</li>
</ul>
<div class="admonition question">
<p class="admonition-title">Exercise: Create some objects in R</p>
<p>Create the following objects; give each object an appropriate name
(your best guess at what name to use is fine):</p>
<ol>
<li>Create an object that has the value of number of pairs of human chromosomes</li>
<li>Create an object that has a value of your favorite gene name</li>
<li>Create an object that has this URL as its value: "<a href="ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/</a>"</li>
<li>Create an object that has the value of the number of chromosomes in a
diploid human cell</li>
</ol>
<details class="success">
<summary>Solution</summary>
<p>Here as some possible answers to the challenge:</p>
<div class="admonition r-project">
<p class="admonition-title">r</p>
</div>
<div class="highlight"><pre><span></span><code><a href="#__codelineno-2-1" id="__codelineno-2-1" name="__codelineno-2-1"></a>human_chr_number &lt;- 23
<a href="#__codelineno-2-2" id="__codelineno-2-2" name="__codelineno-2-2"></a>gene_name &lt;- 'pten'
<a href="#__codelineno-2-3" id="__codelineno-2-3" name="__codelineno-2-3"></a>ensemble_url &lt;- 'ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/'
<a href="#__codelineno-2-4" id="__codelineno-2-4" name="__codelineno-2-4"></a>human_diploid_chr_num &lt;-  2 * human_chr_number
</code></pre></div>
</details>
</div>
<h2 id="naming-objects-in-r">Naming objects in R<a class="headerlink" href="#naming-objects-in-r" title="Permanent link">¶</a></h2>
<p>Here are some important details about naming objects in R</p>
<div class="admonition quote">
<ul>
<li><strong>Avoid spaces and special characters</strong>: Object names cannot contain spaces or the minus sign (<code>-</code>). You can use '_' to make names more readable. You should avoid
  using special characters in your object name (e.g. ! @ # . , etc.). Also,
  object names cannot begin with a number.</li>
<li><strong>Use short, easy-to-understand names</strong>: You should avoid naming your objects
  using single letters (e.g. 'n', 'p', etc.). This is mostly to encourage you
  to use names that would make sense to anyone reading your code (a colleague,
  or even yourself a year from now). Also, avoiding excessively long names will
  make your code more readable.</li>
<li><strong>Avoid commonly used names</strong>: There are several names that may already have a
  definition in the R language (e.g. 'mean', 'min', 'max'). One clue that a name
  already has meaning is that if you start typing a name in RStudio and it gets
  a colored highlight or RStudio gives you a suggested autocompletion you have
  chosen a name that has a reserved meaning.</li>
<li><strong>Use the recommended assignment operator</strong>: In R, we use '&lt;- ' as the
  preferred assignment operator. '=' works too, but is most commonly used in
  passing arguments to functions (more on functions later). There is a shortcut
  for the R assignment operator:</li>
<li>Windows execution shortcut: Alt+-</li>
<li>Mac execution shortcut: Option+-</li>
</ul>
</div>
<p>There are a few more suggestions about naming and style you may want to learn
more about as you write more R code. There are several "style guides" that
have advice, and one to start with is the <a href="http://style.tidyverse.org/index.html">tidyverse R style guide</a>.</p>
<div class="admonition tip">
<p class="admonition-title">Pay attention to warnings in the script console</p>
<p>If you enter a line of code in your script that contains an error, RStudio
may give you an error message and underline this mistake. Sometimes these
messages are easy to understand, but often the messages may need some figuring
out. Paying attention to these warnings will help you avoid mistakes. In the example below, our object name has a space, which
is not allowed in R. The error message does not say this directly,
but R is "not sure"
about how to assign the name to "human_ chr_number" when the object name we
want is "human_chr_number".</p>
<p><a class="glightbox" data-desc-position="bottom" data-height="auto" data-width="100%" href="../figures/rstudio_script_warning.png"><img alt="images" src="../figures/rstudio_script_warning.png" width="700"/></a></p>
</div>
<h2 id="reassigning-object-names-or-deleting-objects">Reassigning object names or deleting objects<a class="headerlink" href="#reassigning-object-names-or-deleting-objects" title="Permanent link">¶</a></h2>
<p>Once an object has a value, you can change that value by overwriting it. R will not give you a warning or error if you overwriting an object, which may or may not be a good thing depending on how you look at it.</p>
<div class="admonition r-project">
<p class="admonition-title">r</p>
<div class="highlight"><pre><span></span><code><a href="#__codelineno-3-1" id="__codelineno-3-1" name="__codelineno-3-1"></a># gene_name has the value 'pten' or whatever value you used in the challenge
<a href="#__codelineno-3-2" id="__codelineno-3-2" name="__codelineno-3-2"></a># We will now assign the new value 'tp53'
<a href="#__codelineno-3-3" id="__codelineno-3-3" name="__codelineno-3-3"></a>
<a href="#__codelineno-3-4" id="__codelineno-3-4" name="__codelineno-3-4"></a>gene_name &lt;- 'tp53'
</code></pre></div>
</div>
<p>You can also remove an object from R's memory entirely. The <code>rm()</code> function will delete the object.</p>
<div class="admonition r-project">
<p class="admonition-title">r</p>
<div class="highlight"><pre><span></span><code><a href="#__codelineno-4-1" id="__codelineno-4-1" name="__codelineno-4-1"></a># delete the object 'gene_name'
<a href="#__codelineno-4-2" id="__codelineno-4-2" name="__codelineno-4-2"></a>rm(gene_name)
</code></pre></div>
</div>
<p>If you run a line of code that has only an object name, R will normally
display the contents of that object. In this case, we are told the
object no longer exists.</p>
<div class="admonition failure">
<p class="admonition-title">Error</p>
<div class="highlight"><pre><span></span><code><a href="#__codelineno-5-1" id="__codelineno-5-1" name="__codelineno-5-1"></a>Error: object 'gene_name' not found
</code></pre></div>
</div>
<h2 id="understanding-object-data-types-modes">Understanding object data types (modes)<a class="headerlink" href="#understanding-object-data-types-modes" title="Permanent link">¶</a></h2>
<p>In R, <strong>every object has two properties</strong>:</p>
<ul>
<li><strong>Length</strong>: How many distinct values are held in that object</li>
<li><strong>Mode</strong>: What is the classification (type) of that object.</li>
</ul>
<p>We will get to the "length" property later in the lesson. The <strong>"mode"
property</strong> <strong>corresponds to the type of data an object represents</strong>. The
most common modes you will encounter in R are:</p>
<hr/>
<p>Mode              Type of data
  (abbreviation)    </p>
<hr/>
<p>Numeric (num)     Numbers such floating point/decimals (1.0, 0.5,
                    3.14), there are also more specific numeric types
                    (dbl - Double, int - Integer). These differences are
                    not relevant for most beginners and pertain to how
                    these values are stored in memory</p>
<p>Character (chr)   A sequence of letters/numbers in single '' or double
                    " " quotes</p>
<p>Logical           Boolean values - TRUE or FALSE</p>
<hr/>
<p>There are a few other modes (i.e. "complex", "raw" etc.) but these are
the three we will work with in this lesson.</p>
<p>Data types are familiar in many programming languages, but also in
natural language where we refer to them as the parts of speech,
e.g. nouns, verbs, adverbs, etc. Once you know if a word - perhaps an
unfamiliar one - is a noun, you can probably guess you can count it and
make it plural if there is more than one (e.g. 1
<a href="https://en.wikipedia.org/wiki/Tuatara">Tuatara</a>, or 2 Tuataras). If
something is a adjective, you can usually change it into an adverb by
adding "-ly"
(e.g. <a href="https://www.merriam-webster.com/dictionary/jejune">jejune</a> vs.
jejunely). Depending on the context, you may need to decide if a word is
in one category or another (e.g "cut" may be a noun when it's on your
finger, or a verb when you are preparing vegetables). These concepts
have important analogies when working with R objects.</p>
<div class="admonition question">
<p class="admonition-title">Exercise: Create objects and check their modes</p>
<p>Create the following objects in R, then use the <code>mode()</code> function to
verify their modes. Try to guess what the mode will be before you look
at the solution</p>
<ol>
<li><code>chromosome_name &lt;- 'chr02'</code></li>
<li><code>od_600_value &lt;- 0.47</code></li>
<li><code>chr_position &lt;- '1001701'</code></li>
<li><code>spock &lt;- TRUE</code></li>
<li><code>pilot &lt;- Earhart</code></li>
</ol>
<details class="success">
<summary>Solution</summary>
<p><code>{r, echo = FALSE, purl = FALSE} chromosome_name &lt;- 'chr02' od_600_value &lt;- 0.47 chr_position &lt;- '1001701' spock &lt;- TRUE pilot &lt;- Earhart</code></p>
<p><code>{r, purl = FALSE} mode(chromosome_name) mode(od_600_value) mode(chr_position) mode(spock) mode(pilot)</code>
{: .solution} {: .challenge}</p>
</details>
</div>
<p>Notice from the solution that even if a series of numbers is given as a
value R will consider them to be in the "character" mode if they are
enclosed as single or double quotes. Also, notice that you cannot take a
string of alphanumeric characters (e.g. Earhart) and assign as a value
for an object. In this case, R looks for an object named <code>Earhart</code> but
since there is no object, no assignment can be made. If <code>Earhart</code> did
exist, then the mode of <code>pilot</code> would be whatever the mode of <code>Earhart</code>
was originally. If we want to create an object called <code>pilot</code> that was
the <strong>name</strong> "Earhart", we need to enclose <code>Earhart</code> in quotation marks.</p>
<div class="admonition r-project">
<p class="admonition-title">r</p>
<p><code>pilot &lt;- "Earhart" mode(pilot)</code></p>
</div>
<h2 id="mathematical-and-functional-operations-on-objects">Mathematical and functional operations on objects<a class="headerlink" href="#mathematical-and-functional-operations-on-objects" title="Permanent link">¶</a></h2>
<p>Once an object exists (which by definition also means it has a mode), R
can appropriately manipulate that object. For example, objects of the
numeric modes can be added, multiplied, divided, etc. R provides several
mathematical (arithmetic) operators including:</p>
<p>Operator     Description</p>
<hr/>
<p>+           addition
  -           subtraction
  *           multiplication
  /            division
  ^ or **   exponentiation
  a%/%b        integer division (division where the remainder is discarded)
  a%%b         modulus (returns the remainder after division)</p>
<p>These can be used with literal numbers:</p>
<p><code>{r, purl = FALSE} (1 + (5 ** 0.5))/2</code></p>
<p>and importantly, can be used on any object that evaluates to
(i.e. interpreted by R) a numeric object:</p>
<p><code>{r, echo = FALSE, purl = FALSE} human_chr_number &lt;- 23</code></p>
<p>```{r, purl = FALSE} # multiply the object 'human_chr_number' by 2</p>
<p>human_chr_number * 2</p>
<div class="highlight"><pre><span></span><code>&gt; ## Exercise: Compute the golden ratio
&gt;
&gt; One approximation of the golden ratio (φ) can be found by taking the sum of 1
&gt; and the square root of 5, and dividing by 2 as in the example above. Compute
&gt; the golden ratio to 3 digits of precision using the `sqrt()` and `round()`
&gt; functions. Hint: remember the `round()` function can take 2 arguments.
&gt;
&gt;&gt; ## Solution
&gt;&gt; ```{r, purl = FALSE}
&gt;&gt; round((1 + sqrt(5))/2, digits = 3)
&gt;&gt; ```
&gt;&gt; Notice that you can place one function inside of another.
&gt; {: .solution}
{: .challenge}

## Vectors

Vectors are probably the
most used commonly used object type in R. 
**A vector is a collection of values that are all of the same type (numbers, characters, etc.)**. 
One of the most common
ways to create a vector is to use the `c()` function - the "concatenate" or
"combine" function. Inside the function you may enter one or more values; for
multiple values, separate each value with a comma:

```{r, purl = FALSE}
# Create the SNP gene name vector

snp_genes &lt;- c("OXTR", "ACTN3", "AR", "OPRM1")
</code></pre></div>
<p>Vectors always have a <strong>mode</strong> and a <strong>length</strong>. You can check these
with the <code>mode()</code> and <code>length()</code> functions respectively. Another useful
function that gives both of these pieces of information is the <code>str()</code>
(structure) function.</p>
<p><code>{r, purl = FALSE} # Check the mode, length, and structure of 'snp_genes' mode(snp_genes) length(snp_genes) str(snp_genes)</code></p>
<p>Vectors are quite important in R. Another data type that we will work
with later in this lesson, data frames, are collections of vectors. What
we learn here about vectors will pay off even more when we start working
with data frames.</p>
<h2 id="creating-and-subsetting-vectors">Creating and subsetting vectors<a class="headerlink" href="#creating-and-subsetting-vectors" title="Permanent link">¶</a></h2>
<p>Let's create a few more vectors to play around with:</p>
<p>```{r, purl = FALSE} # Some interesting human SNPs # while accuracy
is important, typos in the data won't hurt you here</p>
<p>snps \&lt;- c("rs53576", "rs1815739", "rs6152", "rs1799971")
snp_chromosomes \&lt;- c("3", "11", "X", "6") snp_positions \&lt;- c(8762685,
66560624, 67545785, 154039662)</p>
<div class="highlight"><pre><span></span><code>Once we have vectors, one thing we may want to do is specifically retrieve one
or more values from our vector. To do so, we use **bracket notation**. We type
the name of the vector followed by square brackets. In those square brackets
we place the index (e.g. a number) in that bracket as follows:

```{r, purl = FALSE}
# get the 3rd value in the snp vector
snp[3]
</code></pre></div>
<p>In R, every item your vector is indexed, starting from the first item
(1) through to the final number of items in your vector. You can also
retrieve a range of numbers:</p>
<p>```{r, purl = FALSE} # get the 1<sup>st</sup> through 3<sup>rd</sup> value in the snp
vector</p>
<p>snp[1:3]</p>
<div class="highlight"><pre><span></span><code>If you want to retrieve several (but not necessarily sequential) items from
a vector, you pass a **vector of indices**; a vector that has the numbered
positions you wish to retrieve.

```{r, purl = FALSE}
# get the 1st, 3rd, and 4th value in the snp vector

snp[c(1, 3, 4)]
</code></pre></div>
<p>There are additional (and perhaps less commonly used) ways of subsetting
a vector (see <a href="https://thomasleeper.com/Rcourse/Tutorials/vectorindexing.html">these
examples</a>).
Also, several of these subsetting expressions can be combined:</p>
<p><code>{r, purl = FALSE} # get the 1st through the 3rd value, and 4th value in the snp vector # yes, this is a little silly in a vector of only 4 values. snp[c(1:3,4)]</code></p>
<h2 id="adding-to-removing-or-replacing-values-in-existing-vectors">Adding to, removing, or replacing values in existing vectors<a class="headerlink" href="#adding-to-removing-or-replacing-values-in-existing-vectors" title="Permanent link">¶</a></h2>
<p>Once you have an existing vector, you may want to add a new item to it.
To do so, you can use the <code>c()</code> function again to add your new value:</p>
<p><code>{r, purl = FALSE} # add the gene "CYP1A1" and "APOA5" to our list of snp genes # this overwrites our existing vector snp_genes &lt;- c(snp_genes, "CYP1A1", "APOA5")</code></p>
<p>We can verify that "snp_genes" contains the new gene entry</p>
<p><code>{r, purl = FALSE} snp_genes</code></p>
<p>Using a negative index will return a version of a vector with that
index's value removed:</p>
<p><code>{r, purl = FALSE} snp_genes[-6]</code></p>
<p>We can remove that value from our vector by overwriting it with this
expression:</p>
<p><code>{r, purl = FALSE} snp_genes &lt;- snp_genes[-6] snp_genes</code></p>
<p>We can also explicitly rename or add a value to our index using double
bracket notation:</p>
<p><code>{r, purl = FALSE} snp_genes[6]&lt;- "APOA5" snp_genes</code></p>
<blockquote>
<h2 id="exercise-examining-and-subsetting-vectors">Exercise: Examining and subsetting vectors<a class="headerlink" href="#exercise-examining-and-subsetting-vectors" title="Permanent link">¶</a></h2>
<p>Answer the following questions to test your knowledge of vectors</p>
<p>Which of the following are true of vectors in R? A) All vectors have a
mode <strong>or</strong> a length\
B) All vectors have a mode <strong>and</strong> a length\
C) Vectors may have different lengths\
D) Items within a vector may be of different modes\
E) You can use the <code>c()</code> to add one or more items to an existing
vector\
F) You can use the <code>c()</code> to add a vector to an exiting vector\
&gt; &gt; ## Solution &gt; &gt; A) False - Vectors have both of these
properties\
&gt; B) True\
&gt; C) True\
&gt; D) False - Vectors have only one mode (e.g. numeric, character);
all items in\
&gt; a vector must be of this mode. &gt; E) True\
&gt; F) True\
{: .solution} {: .challenge}</p>
</blockquote>
<h2 id="logical-subsetting">Logical Subsetting<a class="headerlink" href="#logical-subsetting" title="Permanent link">¶</a></h2>
<p>There is one last set of cool subsetting capabilities we want to
introduce. It is possible within R to retrieve items in a vector based
on a logical evaluation or numerical comparison. For example, let's say
we wanted get all of the SNPs in our vector of SNP positions that were
greater than 100,000,000. We could index using the '&gt;' (greater than)
logical operator:</p>
<p><code>{r, purl = FALSE} snp_positions[snp_positions &gt; 100000000]</code></p>
<p>In the square brackets you place the name of the vector followed by the
comparison operator and (in this case) a numeric value. Some of the most
common logical operators you will use in R are:</p>
<p>Operator   Description</p>
<hr/>
<p>\&lt;         less than
  \&lt;=        less than or equal to
  &gt;         greater than
  &gt;=        greater than or equal to
  ==         exactly equal to
  !=         not equal to
  !x         not x
  a | b     a or b
  a &amp; b      a and b</p>
<blockquote>
<h2 id="the-magic-of-programming">The magic of programming<a class="headerlink" href="#the-magic-of-programming" title="Permanent link">¶</a></h2>
<p>The reason why the expression
<code>snp_positions[snp_positions &gt; 100000000]</code> works can be better
understood if you examine what the expression "snp_positions &gt;
100000000" evaluates to:</p>
<p><code>{r, purl = FALSE} snp_positions &gt; 100000000</code></p>
<p>The output above is a logical vector, the 4<sup>th</sup> element of which is
TRUE. When you pass a logical vector as an index, R will return the
true values:</p>
<p><code>{r, purl = FALSE} snp_positions[c(FALSE, FALSE, FALSE, TRUE)]</code></p>
<p>If you have never coded before, this type of situation starts to
expose the "magic" of programming. We mentioned before that in the
bracket notation you take your named vector followed by brackets which
contain an index: <strong>named_vector[index]</strong>. The "magic" is that the
index needs to <em>evaluate to</em> a number. So, even if it does not appear
to be an integer (e.g. 1, 2, 3), as long as R can evaluate it, we will
get a result. That our expression
<code>snp_positions[snp_positions &gt; 100000000]</code> evaluates to a number can
be seen in the following situation. If you wanted to know which
<strong>index</strong> (1, 2, 3, or 4) in our vector of SNP positions was the one
that was greater than 100,000,000?</p>
<p>We can use the <code>which()</code> function to return the indices of any item
that evaluates as TRUE in our comparison:</p>
<p><code>{r, purl = FALSE} which(snp_positions &gt; 100000000)</code></p>
<p><strong>Why this is important</strong></p>
<p>Often in programming we will not know what inputs and values will be
used when our code is executed. Rather than put in a pre-determined
value (e.g 100000000) we can use an object that can take on whatever
value we need. So for example:</p>
<p><code>{r, purl = FALSE} snp_marker_cutoff &lt;- 100000000 snp_positions[snp_positions &gt; snp_marker_cutoff]</code></p>
<p>Ultimately, it's putting together flexible, reusable code like this
that gets at the "magic" of programming! {: .callout}</p>
</blockquote>
<h2 id="a-few-final-vector-tricks">A few final vector tricks<a class="headerlink" href="#a-few-final-vector-tricks" title="Permanent link">¶</a></h2>
<p>Finally, there are a few other common retrieve or replace operations you
may want to know about. First, you can check to see if any of the values
of your vector are missing (i.e. are <code>NA</code>, that stands for
<code>not avaliable</code>). Missing data will get a more detailed treatment later,
but the <code>is.NA()</code> function will return a logical vector, with TRUE for
any NA value:</p>
<p>```{r, purl = FALSE} # current value of 'snp_genes': # chr [1:7]
"OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</p>
<p>is.na(snp_genes)</p>
<div class="highlight"><pre><span></span><code>Sometimes, you may wish to find out if a specific value (or several values) is
present a vector. You can do this using the comparison operator `%in%`, which
will return TRUE for any value in your collection that is in
the vector you are searching:

```{r, purl = FALSE}
# current value of 'snp_genes':
# chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"

# test to see if "ACTN3" or "APO5A" is in the snp_genes vector
# if you are looking for more than one value, you must pass this as a vector

c("ACTN3","APOA5") %in% snp_genes
</code></pre></div>
<blockquote>
<h2 id="review-exercise-1">Review Exercise 1<a class="headerlink" href="#review-exercise-1" title="Permanent link">¶</a></h2>
<p>What data modes are the following vectors? a. <code>snps</code>\
b. <code>snp_chromosomes</code>\
c. <code>snp_positions</code></p>
<blockquote>
<h2 id="solution">Solution<a class="headerlink" href="#solution" title="Permanent link">¶</a></h2>
<p><code>{r, purl = FALSE} mode(snps) mode(snp_chromosomes) mode(snp_positions)</code>
{: .solution} {: .challenge}</p>
</blockquote>
<h2 id="review-exercise-2">Review Exercise 2<a class="headerlink" href="#review-exercise-2" title="Permanent link">¶</a></h2>
<p>Add the following values to the specified vectors: a. To the <code>snps</code>
vector add: "rs662799"\
b. To the <code>snp_chromosomes</code> vector add: 11\
c. To the <code>snp_positions</code> vector add: 116792991</p>
<blockquote>
<h2 id="solution_1">Solution<a class="headerlink" href="#solution_1" title="Permanent link">¶</a></h2>
<p><code>{r, purl = FALSE} snps &lt;- c(snps, "rs662799") snps snp_chromosomes &lt;- c(snp_chromosomes, "11") # did you use quotes? snp_chromosomes snp_positions &lt;- c(snp_positions, 116792991) snp_positions</code>
{: .solution} {: .challenge}</p>
</blockquote>
<h2 id="review-exercise-3">Review Exercise 3<a class="headerlink" href="#review-exercise-3" title="Permanent link">¶</a></h2>
<p>Make the following change to the <code>snp_genes</code> vector:</p>
<p>Hint: Your vector should look like this in 'Environment':
<code>chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</code>. If not
recreate the vector by running this expression:
<code>snp_genes &lt;- c("OXTR", "ACTN3", "AR", "OPRM1", "CYP1A1", NA, "APOA5")</code></p>
<p>a.  Create a new version of <code>snp_genes</code> that does not contain CYP1A1
    and then\
b.  Add 2 NA values to the end of <code>snp_genes</code></p>
<blockquote>
<h2 id="solution_2">Solution<a class="headerlink" href="#solution_2" title="Permanent link">¶</a></h2>
<p><code>{r, purl = FALSE} snp_genes &lt;- snp_genes[-5] snp_genes &lt;- c(snp_genes, NA, NA) snp_genes</code>
{: .solution} {: .challenge}</p>
</blockquote>
<h2 id="review-exercise-4">Review Exercise 4<a class="headerlink" href="#review-exercise-4" title="Permanent link">¶</a></h2>
<p>Using indexing, create a new vector named <code>combined</code> that contains: -
The the 1<sup>st</sup> value in <code>snp_genes</code> - The 1<sup>st</sup> value in <code>snps</code> - The 1<sup>st</sup>
value in <code>snp_chromosomes</code> - The 1<sup>st</sup> value in <code>snp_positions</code></p>
<blockquote>
<h2 id="solution_3">Solution<a class="headerlink" href="#solution_3" title="Permanent link">¶</a></h2>
<p><code>{r, purl = FALSE} combined &lt;- c(snp_genes[1], snps[1], snp_chromosomes[1], snp_positions[1]) combined</code>
{: .solution} {: .challenge}</p>
</blockquote>
<h2 id="review-exercise-5">Review Exercise 5<a class="headerlink" href="#review-exercise-5" title="Permanent link">¶</a></h2>
<p>What type of data is <code>combined</code>?</p>
<blockquote>
<h2 id="solution_4">Solution<a class="headerlink" href="#solution_4" title="Permanent link">¶</a></h2>
<p><code>{r, purl = FALSE} typeof(combined)</code> {: .solution} {: .challenge}</p>
</blockquote>
</blockquote>
<h2 id="bonus-material-lists">Bonus material: Lists<a class="headerlink" href="#bonus-material-lists" title="Permanent link">¶</a></h2>
<p>Lists are quite useful in R, but we won't be using them in the genomics
lessons. That said, you may come across lists in the way that some
bioinformatics programs may store and/or return data to you. One of the
key attributes of a list is that, unlike a vector, a list may contain
data of more than one mode. Learn more about creating and using lists
using this <a href="https://r4ds.had.co.nz/vectors.html#lists">nice tutorial</a>.
In this one example, we will create a named list and show you how to
retrieve items from the list.</p>
<p><code>{r, purl = FALSE} # Create a named list using the 'list' function and our SNP examples # Note, for easy reading we have placed each item in the list on a separate line # Nothing special about this, you can do this for any multiline commands # To run this command, make sure the entire command (all 4 lines) are highlighted # before running # Note also, as we are doing all this inside the list() function use of the # '=' sign is good style snp_data &lt;- list(genes = snp_genes,                  refference_snp = snps,                  chromosome = snp_chromosomes,                  position = snp_positions) # Examine the structure of the list str(snp_data)</code></p>
<p>To get all the values for the <code>position</code> object in the list, we use the
<code>$</code> notation:</p>
<p>```{r, purl = FALSE} # return all the values of position object</p>
<p>snp_data$position</p>
<div class="highlight"><pre><span></span><code>To get the first value in the `position` object, use the `[]` notation to index:

```{r, purl = FALSE}
# return first value of the position object

snp_data$position[1]
</code></pre></div>
</article>
</div>
</div>
<a class="md-top md-icon" data-md-component="top" hidden="" href="#">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"></path></svg>
            Back to top
          </a>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top"], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../assets/javascripts/bundle.2a6f1dda.min.js"></script>
<script src="../javascripts/mathjax.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>